File: Desafio3/app/src/main/java/com/example/desafio3/MainActivity.java
package com.example.desafio3;

import androidx.appcompat.app.AppCompatActivity;

import android.net.Uri;
import android.os.Bundle;
import android.widget.Toast;

import com.example.desafio3.api.Api;
import com.example.desafio3.api.RetrofitClient;
import com.example.desafio3.beans.Pregunta;
import com.example.desafio3.beans.PreguntasLista;
import com.example.desafio3.fragments.Fragmento;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "LogGSON";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Api service = RetrofitClient.getRetrofitInstance().create(Api.class);
        Call<PreguntasLista> call = service.getQuestion();

        call.enqueue(new Callback<PreguntasLista>() {
            @Override
            public void onResponse(Call<PreguntasLista> call, Response<PreguntasLista> response) {
                PreguntasLista preguntas = response.body();
                if (preguntas != null) {
                    Pregunta pregunta;
                    pregunta =  preguntas.getResults().get(0);

                    Fragmento frag = null;

                    try {
                        frag = Fragmento.newInstance(
                                URLDecoder.decode(pregunta.getQuestion(), "UTF-8"),
                                URLDecoder.decode(pregunta.getCategory(), "UTF-8"),
                                URLDecoder.decode(pregunta.getDifficulty(), "UTF-8"),
                                URLDecoder.decode(pregunta.getCorrect_answer(), "UTF-8")
                        );
                    } catch (UnsupportedEncodingException e) {
                        e.printStackTrace();
                    }

                    getSupportFragmentManager()
                            .beginTransaction()
                            .add(R.id.fragment_container, frag, "FRAGMENTO")
                            .commit();
                }
            }

            @Override
            public void onFailure(Call<PreguntasLista> call, Throwable t) {
                Toast.makeText(getApplicationContext(), getString(R.string.error_gson), Toast.LENGTH_SHORT);
            }
        });
    }
}

File: Kotlin11/app/src/main/java/com/example/restapi/MainActivity.kt
package com.example.restapi

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.example.restapi.pojo.Post
import com.example.restapi.remote.RetrofitClient
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Response
import retrofit2.Callback

class MainActivity : AppCompatActivity() {

    private var postsList =  ArrayList<Post>()
    private lateinit var viewAdapter: RecyclerView.Adapter<*>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        viewAdapter = PostAdapter(postsList)
        postsRecyclerView.adapter = viewAdapter

        loadApiData()
    }

    private fun loadApiData() {
        val service = RetrofitClient.retrofitInstance()
        val call = service.getAllPosts()

        call.enqueue(object : Callback<ArrayList<Post>> {
            override fun onResponse(
                call: Call<ArrayList<Post>>,
                response: Response<ArrayList<Post>>
            ) {
                response.body()?.map {
                    Log.d("MAIN", "${it.id} - ${it.title}")
                    postsList.add(it)
                }
                viewAdapter.notifyDataSetChanged()
            }

            override fun onFailure(call: Call<ArrayList<Post>>, t: Throwable) {
                Log.d("MAIN", "Error: " + t)
                Toast.makeText(
                    applicationContext,
                    "Error: no pudimos recuperar los posts desde el api",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }
}
